sam.c

```diff
@@ -1924,13 +1938,655 @@ int sam_parse1(kstring_t *s, sam_hdr_t *h, bam1_t *b)
+                    if (bgzf_idx_push(fp->fp.bgzf, fp->idx,
+                                      b->core.tid, b->core.pos, bam_endpos(b),
+                                      bgzf_tell(fp->fp.bgzf),
+                                      !(b->core.flag&BAM_FUNMAP)) < 0) {
+                        sam_state_err(fd, EIO);
+                        goto err;
+                    }
+                } else {
+                    if (hts_idx_push(fp->idx, b->core.tid, b->core.pos, bam_endpos(b),
+                                     bgzf_tell(fp->fp.bgzf), !(b->core.flag&BAM_FUNMAP)) < 0) {
+                        sam_state_err(fd, EIO);
+                        goto err;
+                    }
+                }
+            }
+
+            assert(count == gb->nbams);
+
+            // Add bam array to free-list
+            pthread_mutex_lock(&fd->lines_m);
+            gb->next = fd->bams;
+            fd->bams = gl->bams;
+            gl->bams = NULL;
+            pthread_mutex_unlock(&fd->lines_m);
+        } else {
+            if (fp->format.compression == bgzf) {
+                if (bgzf_write(fp->fp.bgzf, gl->data, gl->data_size) != gl->data_size) {
+                    sam_state_err(fd, EIO);
+                    goto err;
+                }
+            } else {
+                if (hwrite(fp->fp.hfile, gl->data, gl->data_size) != gl->data_size) {
+                    sam_state_err(fd, EIO);
+                    goto err;
+                }
+            }
+        }
+
+        hts_tpool_delete_result(r, 0);
+
+        // Also updated by main thread
+        pthread_mutex_lock(&fd->lines_m);
+        gl->next = fd->lines;
+        fd->lines = gl;
+        pthread_mutex_unlock(&fd->lines_m);
+    }
+
+    sam_state_err(fd, 0); // success
+    hts_tpool_process_destroy(fd->q);
+    fd->q = NULL;
+    return NULL;
+
+ err:
+    return (void *)-1;
+}
+
+// Run from one of the worker threads.
+// Convert a passed in array of BAMs (sp_bams) and converts to a block
+// of text SAM records (sp_lines).
+static void *sam_format_worker(void *arg) {
+    sp_bams *gb = (sp_bams *)arg;
+    sp_lines *gl = NULL;
+    int i;
+    SAM_state *fd = gb->fd;
+    htsFile *fp = fd->fp;
+
+    // Use a block of SAM strings we had earlier if available.
+    pthread_mutex_lock(&fd->lines_m);
+    if (fd->lines) {
+        gl = fd->lines;
+        fd->lines = gl->next;
+    }
+    pthread_mutex_unlock(&fd->lines_m);
+
+    if (gl == NULL) {
+        gl = calloc(1, sizeof(*gl));
+        if (!gl) {
+            sam_state_err(fd, ENOMEM);
+            return NULL;
+        }
+        gl->alloc = gl->data_size = 0;
+        gl->data = NULL;
+    }
+    gl->serial = gb->serial;
+    gl->next = NULL;
+
+    kstring_t ks = {0, gl->alloc, gl->data};
+
+    for (i = 0; i < gb->nbams; i++) {
+        if (sam_format1_append(fd->h, &gb->bams[i], &ks) < 0) {
+            sam_state_err(fd, EIO);
+            return NULL;
```

Comment:
Similar problem here - something needs to take care of `gl` and `gb` to prevent them from leaking.